fix(orc8r): Stabilize cloud tests

MoritzThomasHuebner opened the pull request:
Closes #13998

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>
<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary

Upon inspecting the recent cloud-workflow failures, we found the following flaky tests:

- `TestSqlConfiguratorStorage_LoadEntities`: 2 failures
- `TestGetSubscriberState`: 2 failures
- `TestSingletonRunSuccess`: 6 failures
- `TestSingletonRunFail`: 4 failures
- `TestSyncSubscribers`: 1 failure
- `TestRunBrokenIndexer`: 1 failures

This PR focuses on `TestSingletonRunSuccess`, `TestSingletonRunFail`, and `TestRunBrokenIndexer`.

These changes focus on test isolation:
- Added manual calls to `indexer.DeregisterAllForTest(t)` at the end of the tests. This is not strictly necessary to make the tests green, but it can cause issues with dangling indexers if the order of the tests is ever reversed.

`TestSingletonRunFail`:
- Moved the creation of the reindexer thread after the registration of the indexers. This way, it is impossible that the reindexer is triggered when only one or two out of the three indexers are registered.

`TestSingletonRunSuccess`:
- The issue is that the `TestHookReindexDone` is called even if there is an error during reindexing. Unexpected errors can be triggered in the [remote indexer getIndexerClient](https://github.com/magma/magma/blob/04facb1e03bf532d45827c9df7d38877db299e8b/orc8r/cloud/go/services/state/indexer/remote_indexer.go#L138). 
- Added a `TestHookReindexFailure`, that we use to keep track of reindexing job errors. 
- Safely pull all elements out of the channel `ch` using `recvChSafe`. This function keeps track of failures and will pull the requisite number of elements from the channel.

`TestRunBrokenIndexer`:
- The issue is that in rare cases the reindexer can run twice before the indexer is registered.
- Moved the creation of the reindexer thread after the registration of the indexer, similar to what we do in `TestSingletonRunFail`.
- Removed one `recvCh` call as this appears to be unnecessary now,

## Test Plan
`TestSingletonRunSuccess`:
- Reproduced error locally by manually adding an error that would trigger a fixed number of times in the [remote indexer getIndexerClient](https://github.com/magma/magma/blob/04facb1e03bf532d45827c9df7d38877db299e8b/orc8r/cloud/go/services/state/indexer/remote_indexer.go#L138). 
- The fix in this PR fixes the error that occurs when we manually insert an error.
- It was not possible to recreate this issue on the CI, since copy/pasting the test did not appear to increase the likelihood compared to a single run.

`TestSingletonRunFail`
- Reproduced error by re-running many times on the CI initially. 
- After adding the change, the test is green even with many re-runs:  [workflow](https://github.com/magma/magma/actions/runs/3201345383).

`TestRunBrokenIndexer`:
- Error output can be reproduced locally by deleting one of the `recvCh` calls in the function.
- After adding the change, the test is green locally.
- Unfortunately, the test can not be re-run on the github runner many times in a single workflow as it fails for a different reason when running the second time.

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

MoritzThomasHuebner committed:
fix(orc8r): Stabilize reindexer singleton tests

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Reverted all temporary changes

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Added Connection failure test hook

Keeps track of connection failures when running tests.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Added safe checks for reindexing failures

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Some refactoring to existing changes

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Simplified some changes

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Stabilized TestRunBrokenIndexer

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Changed += 1 to ++

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

MoritzThomasHuebner committed:
fix(orc8r): Improved `TestRunBrokenIndexer` stability

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>

sebathomas commented on orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go:
```diff
@@ -105,22 +112,50 @@ func TestSingletonRunSuccess(t *testing.T) {
+	require.Equal(t, 3+reindexConnectionFailsTotal, reindexDoneNum)
+}
+
+func recvChSafe(t *testing.T, ch chan interface{}) int {
```
Maybe a better name would make this more readable. For example `recvChAndRetryFailures`?

Also it would be good to document that we return the number of failed runs.

sebathomas commented on orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go:
```diff
@@ -105,22 +112,50 @@ func TestSingletonRunSuccess(t *testing.T) {
+	}
+	defer func() { reindex.TestHookReindexFailure = func() {} }()
+	remainingNumberOfReindexingRuns := 1
+	for {
```
Is there a global test timeout so the CI won't get stuck in an endless loop if the reindexer keeps failing? Might be good to add a reasonable number of maximum retries.

sebathomas reviewed (commented):
This didn't exactly make the test easier to understand. ;) I think it's looking good but there's a few changes I'm not sure about. Maybe let's go over it tomorrow in person.

MoritzThomasHuebner committed:
fix(orc8r): Some changes for the review

- Renamed `recvChSafe` to `recvChAndRetryFailures`.
- Added global check on `reindexConnectionFailsTotal` so it aborts if the connection keeps failing.

Signed-off-by: Moritz Huebner <moritz.huebner@tngtech.com>
